### PR TITLE
Backend Logic for new Emails/Notifications

### PR DESCRIPTION
**Pull request by stkem**
@andrewconner

### CODE REVIEW

**Comment by stkem** [comment]
Need to clean up the welcome email stuff in particular.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala [hunk 1]
```diff
@@ -164,4 +172,80 @@ class UserCommander @Inject() (
 
     newUser
   }
+
+  def tellAllFriendsAboutNewUser(newUserId: Id[User], additionalRecipients: Seq[Id[User]]): Unit = {
```
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala:197]: this is not a trivial piece of code, please add tests. the `TestMailModule` with `FakeOutbox` makes it very easy to check what emails where supposed to be sent.

File: kifi-backend/modules/common/app/com/keepit/search/message/ThreadContent.scala [hunk 2]
```diff
@@ -12,8 +12,8 @@ import play.api.libs.functional.syntax._
 import org.joda.time.DateTime
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/search/message/ThreadContent.scala:12]: This file is unrelated (it's part of message search). Slipped in by accident.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala [hunk 3]
```diff
@@ -165,6 +178,90 @@ class UserCommander @Inject() (
...
+            to = List(id2Email(userId)),
+            subject = s"${newUser.firstName} ${newUser.lastName} joined KiFi!",
+            htmlBody = views.html.email.invitationFriendJoined(newUser).body,
+            category = PostOffice.Categories.User.INVITATION)
+          )
+        }
+
+      }
+
+      elizaServiceClient.sendGlobalNotification( //ZZZ update this with correct copy, etc.
```
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala]: This (`sendGlobalNotification`) needs a better name. Suggestions welcome.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala [hunk 4]
```diff
@@ -165,6 +178,90 @@ class UserCommander @Inject() (
     newUser
   }
 
+  def tellAllFriendsAboutNewUser(newUserId: Id[User], additionalRecipients: Seq[Id[User]]): Unit = {
+    val guardKey = "friendsNotifiedAboutJoining"
+    if (!db.readOnly{ implicit session => userValueRepo.getValue(newUserId, guardKey).exists(_=="true") }) {
```
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala:199]: Did you consider using a key with the opposite meaning, persisted when the user is created, then cleared when the notifications are sent?

Potential advantages:
- Would give us the freedom to use whatever trigger we want to send the notifications (e.g. a cron job), instead of relying on the code path associated with a particular user action
- Could make it easier to identify users for which notifications should go out (a bit easier to distinguish old users like you and me from new users)

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala [hunk 5]
```diff
@@ -165,6 +178,90 @@ class UserCommander @Inject() (
...
+      val (newUser, toNotify, id2Email) = db.readOnly { implicit session =>
+        val newUser = userRepo.get(newUserId)
+        val toNotify = userConnectionRepo.getConnectedUsers(newUserId) ++ additionalRecipients //ZZZ does this include social connections
+        val id2Email = toNotify.map { userId =>
+          (userId, emailRepo.getByUser(userId))
+        }.toMap
+        (newUser, toNotify, id2Email)
+      }
+      toNotify.foreach { userId => //ZZZ update content here to the correct content
+        db.readWrite{ implicit session =>
```
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala:212]: Why a session for each email? Seems safer to use one session and to update `userValueRepo` in the same session. (Granted, I don’t know much about our email-sending infrastructure.)

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala (hunk 4 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala:199]: That's a good point. Unless we pre-set all the current users, we'll be sending a lot of emails on behalf of people who have been members for a while.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala (hunk 5 above)
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala:212]: If a long write session is the concern, then we could do a bulk insert into some other table to create rows that persistently track which users should get emails and which have. Or maybe we don’t care too much about integrity/completeness of these notifications yet.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala (hunk 3 above)
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala]: I would prefer using a different notification type for automated notifications, to distinguish them from manual announcements from our team that are generally intended for all users. Right now `"global"` notifications are labeled as an “Announcement” in the extension. Two possible type names you might consider are `"system"` and `"automated"`.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/UserController.scala [hunk 6]
```diff
@@ -155,9 +157,63 @@ class UserController @Inject() (
...
+                  senderUserId = None,
+                  from = EmailAddresses.INVITATION,
+                  fromName = Some(s"${requestingUser.firstName} ${requestingUser.lastName}"),
+                  to = List(destinationEmail),
+                  subject = s"${requestingUser.firstName} ${requestingUser.lastName} wants to be kifi friends with you!", //ZZZ plug into correct copy and template when ready
+                  htmlBody = s"${requestingUser.firstName} ${requestingUser.lastName} wants to be kifi friends with you!",
+                  category = PostOffice.Categories.User.INVITATION)
+                )(session)
+              }
+              elizaServiceClient.sendGlobalNotification( //ZZZ update this with correct copy, etc.
```
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/UserController.scala]: See earlier comment. I don’t think these new automated, personalized notifications should fall under the same umbrella (`"global"`) as our manually sent, broadcasted notifications.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala (hunk 4 above)
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala:199]: Sorry, somehow missed this earlier. Very good point! I'll make the necessary change. We don't want to spam out existing users.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala (hunk 3 above)
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/UserCommander.scala]: Sure, very easy change on the backend as the actual logic is still basically the same. I tried to keep it so that it would be transparent to the extension (briefly discussed with product).
We discuss how to best structure it when we are both back in the office.
